### ISSUE TITLE
`#[clippy::format_args]` does not support macros that adapt arguments

### ISSUE THREAD

**Issue by tamird** [rust-lang/rust-clippy#16833]
### Summary

In Rust-for-Linux we use a procedural macro to apply a rewrite before delegating to `core::format_args!`; we wrap all arguments in `Adapter` which allows us to implement `Display` for types that do not implement it in `core` e.g. [`core::ffi::CStr`](https://doc.rust-lang.org/stable/core/ffi/struct.CStr.html).

It turns out that annotated macros which internally apply this rewrite are somehow prevented from triggering `uninlined_format_args`. **Note that the repro I've provided is not perfect; rewriting to inline format args there would produce broken code, but the [actual macro in the kernel](https://github.com/torvalds/linux/blob/a55f7f5f29b32c2c53cc291899cf9b0c25a07f7c/rust/macros/fmt.rs#L8-L96) is smart enough to rewrite `foo!("{i}")` into `format_args!("{i}", i = Adapter(&i))`.**

Please note there are good reasons for both decisions mentioned at the top: implementing `Display` implies being able to round trip through `ToString` and `FromStr`, but in the kernel the ergonomics of having `Display` on `CStr` are important. Let's refrain from litigating these decisions here.

### Lint Name

uninlined_format_args

### Reproducer

I tried this code [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=7c07a26c854499011002bfcb2bdc0e9b):

```rust
#![warn(clippy::uninlined_format_args)]

use std::fmt;

struct Adapter<T>(T);

impl<T: fmt::Display> fmt::Display for Adapter<&T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.0.fmt(f)
    }
}

#[clippy::format_args]
macro_rules! forwarding_format_args {
    ($($args:tt)*) => {
        format_args!($($args)*)
    };
}

#[clippy::format_args]
macro_rules! adapted_format_args {
    ($fmt:literal, $arg:expr) => {
        format_args!($fmt, Adapter(&($arg)))
    };
}

#[clippy::format_args]
macro_rules! forwarding_adapted_format_args {
    ($($args:tt)*) => {
        adapted_format_args!($($args)*)
    };
}

pub fn demo(x: i32) {
    #[expect(clippy::uninlined_format_args)]
    let _ = format_args!("{}", x);
    #[expect(clippy::uninlined_format_args)]
    let _ = forwarding_format_args!("{}", x);

    // Expected: `#[clippy::format_args]` should make this equivalent to a
    // formatting macro invocation and warn that `x` can be inlined.
    //
    // Actual: Clippy does not warn because the macro expansion wraps `x` in
    // `Adapter(&x)` before the eventual `format_args!` invocation.
    #[expect(clippy::uninlined_format_args)]
    let _ = adapted_format_args!("{}", x);
    #[expect(clippy::uninlined_format_args)]
    let _ = forwarding_adapted_format_args!("{}", x);
}
```

I expected to see this happen:

No warnings.

Instead, this happened:

```
    Checking playground v0.0.1 (/playground)
warning: this lint expectation is unfulfilled
  --> src/lib.rs:45:14
   |
45 |     #[expect(clippy::uninlined_format_args)]
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(unfulfilled_lint_expectations)]` on by default

warning: this lint expectation is unfulfilled
  --> src/lib.rs:47:14
   |
47 |     #[expect(clippy::uninlined_format_args)]
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: `playground` (lib) generated 2 warnings
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.61s
```

### Version

```text

```

**Comment by Souradip121** [comment]
@rustbot claim

Approach:

There were two separate bugs preventing the lint from firing:

1. FormatArgsCollector (format_args_collector.rs) — has_span_from_proc_macro false positive

This function checks whether a FormatArgs node should be skipped (because its spans were set by a proc macro). It works by checking the "between" spans — the source text gaps between the format string and each argument.

For adapted macros, walk_chain(Adapter(&(x)).span, call_site_ctxt) resolves to the entire outer macro call span, which starts before the format string literal. This makes start.between(end) produce an inverted/invalid span. The original code passed this to check_source_text, which returned false, causing has_span_from_proc_macro to return true — wrongly treating the FormatArgs as proc-macro-generated and refusing to store it. Without storage, the late lint pass never sees it.

Fix: When the between-span is inverted (sp.lo() < start.hi()), skip the check and treat it as valid — but only when !args.span.from_expansion() (i.e., the format string is at the call site, not generated by something like concat!).

2. check_one_arg (format_args.rs) — only handled bare ExprKind::Path arguments

The existing code required the format argument to be a simple bare path (e.g. x). For adapted macros, the format_args! inside the macro body receives Adapter(&(x)) — an ExprKind::Call — so the pattern match failed and no fix was generated.

Fix: Added a find_callsite_simple_path helper that recurses through macro-body wrappers (Call, AddrOf, Paren, Unary, Cast) to find the single call-site variable that was substituted as $arg. The function returns None if the call-site expression isn't a simple path (e.g. x + 1) — preventing false positives. The removal span for the argument is computed using the call-site variable span directly, anchored to the end of the previous item.
